libarchive/archive_write_open_filename.c

```diff
@@ -64,9 +64,28 @@ struct write_file_data {
 static int	file_close(struct archive *, void *);
 static int	file_free(struct archive *, void *);
 static int	file_open(struct archive *, void *);
+static int	file_open_fd(struct archive *, void *);
 static ssize_t	file_write(struct archive *, void *, const void *buff, size_t);
 static int	open_filename(struct archive *, int, const void *);
 
+int
+archive_write_open_fd(struct archive *a, int fd)
```

Comment:
I don't have a strong opinion about merging `archive_write_open_fd.c` and `archive_write_open_filename.c`.  There are a couple of guiding principles at work:
* Libarchive tries hard to keep everything `static` that isn't part of the official API.  That avoids the need to fuss with linker-specific tricks to manage symbol visibility, but also implies that we try to limit our use of "private" internal interfaces.
* Libarchive also tries hard to avoid static link pollution:  If you reference a symbol, you get just the necessary pieces needed to support that symbol and very little else.  But that means we have to be careful about combining features into a single source file.

A side effect of the above principles is that we do naturally get some duplication of common basic functionality.  Personally, I'm fine with that:  it has the side effect of keeping most source files standalone and easy to understand without having to dig through the entire system to untangle a lot of specialized vernacular.  It means that most library functionality could be implemented outside of the library without any problems.

In this particular case, I don't think there's a really compelling argument either way:  The duplicated functions aren't that significant and neither is the link pollution from combining the two.  But I would not like to see other unrelated files get merged just because they happen to share a few helper functions.